Add reference widgets to legacy widget blocks. Use ajax admin endpoint on this blocks.

jorgefilipecosta opened the pull request:
## Description
Fixes: https://github.com/WordPress/gutenberg/issues/16604

This is a WIP it's working correctly but I still need to check the code to improve it.

During the widgets RFC, we decided that existing widgets will be referenced in the legacy widgets and the real instance should be updated in these cases.
This PR uses the ajax admin endpoint to implement this concept of referencing widgets in the legacy widgets block.
This mechanism is also used to implement callback widgets so this PR supersedes https://github.com/WordPress/gutenberg/pull/14395 and allows us to avoid the hack `$_POST   = array_merge( $_POST, $instance_changes );` present in the other PR.
The rest endpoint is used to get the initial form of all widgets (the ajax admin does not seem to allow that), and it is used to provide the update logic in class widgets that don't have an instance. That endpoint should be read-only unless the widgets apply changes in methods that should not have side-effects.
The ajax admin is used to perform updates on legacy widgets that reference a real widget instance present in the database.

## How has this been tested?
I pasted the following code in the block editor:
```
<!-- wp:legacy-widget {"widgetClass":"WP_Widget_Archives","identifier":"archives-12"} /-->

<!-- wp:legacy-widget {"widgetClass":"WP_Widget_Archives","identifier":"archives-11"} /-->

<!-- wp:legacy-widget {"widgetClass":"WP_Widget_Media_Audio","identifier":"media_audio-26"} /-->
```
(Update identifier to real widget instances that were create using the widgets screen).

I verified the legacy widget block works as expected and if I open the widgets screen the changes I did appear there.

I used "NattyWP feedburner widget". The widget comes with silesia theme that can be downloaded at wordpress.org/themes/silesia. The theme is not updated to WordPress 5.0 and some warning messages appear. For testing purposes, the widget can also be used standalone by pasting the code available themes.trac.wordpress.org/browser/silesia/1.0.6/include/widgets/feedburner.php in the functions.php of any other theme.

jorgefilipecosta committed:
Fix: End to end tests do not disable the experiments after they are enabled.

noisysocks commented:
> This is a WIP it's working correctly but I still need to check the code to improve it.

Is it still a WIP?

jorgefilipecosta commented:
> Is it still a WIP?

It suffered a huge rebase but is ready for a review.
With these changes, the legacy widgets should load and allow edits on the widget screen. Preview is not working because it is pending on https://github.com/WordPress/gutenberg/pull/15635.

jorgefilipecosta commented:
This PR was rebased. In my tests, it worked well after the rebase.
It fixes issue https://github.com/WordPress/gutenberg/issues/16604, so we should get it in.
@noisysocks do you have some availability to review the PR?

noisysocks commented:
Really really cool to see the Legacy Widget block working and doing stuff!

I'm a little confused, how does this fix #16604? When I don't have any plugins installed and insert a Legacy Widget block, I still see an empty dropdown:

<img width="646" alt="Screen Shot 2019-08-30 at 16 30 18" src="https://user-images.githubusercontent.com/612155/63998011-78b6bd00-cb43-11e9-8608-f4754821a08b.png">

---

It's really apparent that we need to show the title of the widget in the block. For example, you can't tell at all that this is a Search widget:

<img width="652" alt="Screen Shot 2019-08-30 at 16 31 15" src="https://user-images.githubusercontent.com/612155/63998112-bb789500-cb43-11e9-8e91-4f56c4d4e43c.png">

`widgets.php` solves this by having a header with the title in it:

<img width="456" alt="Screen Shot 2019-08-30 at 16 33 26" src="https://user-images.githubusercontent.com/612155/63998208-f11d7e00-cb43-11e9-86f5-5a6fe34e8a3a.png">

---

I tried testing this PR with [Image Widget](https://wordpress.org/plugins/image-widget/), which was the first plugin I found when searching the directory for "widget".

I was able to insert a Legacy Widget block and select _Image Widget_, but then I was not able to interact with the widget at all. Clicking on _Select Image_ just caused the page to reload.

I had similar problem with other third party widget plugins that I tested.

jorgefilipecosta commented:
Hi @noisysocks,

> I'm a little confused, how does this fix #16604? When I don't have any plugins installed and insert a Legacy Widget block, I still see an empty dropdown:

Previously we were showing "Block Areas" widget in legacy widgets. This PR fixes that. The empty dropdown was a bug and not it shows as expected a message saying there is no widget available.

> I tried testing this PR with Image Widget, which was the first plugin I found when searching the directory for "widget".

> I was able to insert a Legacy Widget block and select Image Widget, but then I was not able to interact with the widget at all. Clicking on Select Image just caused the page to reload.

There are some known problems where some third-party widgets don't work because of relying on a specific dom.
A good way to test legacy widgets is enabling the core widgets by setting the array set on https://github.com/WordPress/gutenberg//blob/a4ceeb0159b6dab27a68c122c9c86c4559258b87/lib/widgets.php#L94 to empty.
Core Widgets should work as expected as they rely on the expected widget events.


That said the Image widget seems to work on my tests:
![Sep-03-2019 14-33-14](https://user-images.githubusercontent.com/11271197/64178038-93957400-ce58-11e9-8b0f-7c8610803dad.gif)

Is there any error being logged during your tests?

jorgefilipecosta commented:
> It's really apparent that we need to show the title of the widget in the block. For example, you can't tell at all that this is a Search widget:

We show the currently selected widget in the inspector:
![image](https://user-images.githubusercontent.com/11271197/64178136-cd667a80-ce58-11e9-868e-e3d49c3dfd70.png)

We are trying to follow what's happening in other blocks. If we have multiple paragraph blocks, it is not easy to tell which type of paragraph a block is. The user would only know by selecting the block and checking its name on the inspector.
But on widgets, the problem is bigger as most widgets look the same.
cc: @mapk

mapk commented:
> It's really apparent that we need to show the title of the widget in the block.

I'm torn on this one. I'd like to stay inline with how we treat other blocks. But in this case, it's not a block, it's a widget _inside_ a block. Can we adopt the reusable block pattern of displaying a name? Maybe we don't include the "Edit" button in this case though.

![Screen Shot 2019-09-06 at 2 20 06 PM](https://user-images.githubusercontent.com/617986/64461238-73acbd00-d0b1-11e9-9000-5bc3e137859d.png)

jorgefilipecosta commented:
Hi @noisysocks, @mapk,
The design of legacy widgets was updated and now it contains a title indication similar to reusable blocks.
<img width="875" alt="Screenshot 2019-10-24 at 17 58 48" src="https://user-images.githubusercontent.com/11271197/67509175-5300ec80-f68a-11e9-8611-316291544ed2.png">
<img width="881" alt="Screenshot 2019-10-24 at 18 02 09" src="https://user-images.githubusercontent.com/11271197/67509177-5300ec80-f68a-11e9-9c24-272a8e41688c.png">

noisysocks commented:
A few issues I noticed while testing:

1. The Legacy Widget block needs need a white background so that it doesn't look strange when the editor CSS has a non-white background:
   ![localhost_8889_wp-admin_post-new php](https://user-images.githubusercontent.com/612155/67651663-328ca700-f996-11e9-87c4-37d627f564b1.png)
2. It's nice that we show the name of the widget now, but because we only do this when the Legacy Widget block is selected, things look pretty confusing when you first land at the Block Areas page:
   ![localhost_8889_wp-admin_admin php_page=gutenberg-widgets](https://user-images.githubusercontent.com/612155/67652061-7af89480-f997-11e9-8900-6cc1aa2dac58.png)
3. When I make a change to a Legacy Widget block in the Block Areas page, the _Update_ button in the top right remains disabled.

noisysocks commented on lib/load.php:
```diff
@@ -20,6 +20,8 @@
 	if ( ! class_exists( 'WP_REST_Widget_Areas_Controller' ) ) {
 		require dirname( __FILE__ ) . '/class-experimental-wp-widget-blocks-manager.php';
+	}
+	if ( ! class_exists( 'WP_REST_Widget_Areas_Controller' ) ) {
```
These two `if`s are the same.

noisysocks commented on lib/widgets.php:
```diff
@@ -63,6 +63,14 @@ function gutenberg_block_editor_admin_print_footer_scripts() {
+				wp_nonce_field( 'save-sidebar-widgets', '_wpnonce_widgets', false ),
+				'</form>',
+			)
+		);
```
What's this for?

jorgefilipecosta replied:
I added a comment. Basically for the widgets code to be able to call the ajax endpoint we need a nonce.

noisysocks commented on lib/widgets.php:
```diff
@@ -124,10 +132,12 @@ function gutenberg_get_legacy_widget_settings() {
 				isset( $widget_obj['callback'][0] ) &&
-				( $widget_obj['callback'][0] instanceof WP_Widget )
+				( $widget_obj['callback'][0] instanceof WP_Widget ) ) ||
+				strncmp( $widget_id, $block_widget_start, strlen( $block_widget_start ) ) === 0
```
It's more common (and, I think, easier to understand) to see `substr` used to check that a string begins with another string.

```php
substr( $widget_id, 0, strlen( $block_widget_start ) ) === $block_widget_start
```

jorgefilipecosta replied:
Performance is not very relevant in this case. But, substring first unconditionally iterates on the string to copy parts of it and the only in the end the comparison is done. The method I'm using does not copy any character and stops the comparison when the first char difference is found. So I normally use this code because it is more performant. I added a comment to say that this logic is a starts with.

noisysocks replied:
Interesting! PHP needs a `starts_with` function 😛

noisysocks commented on packages/block-library/src/legacy-widget/index.php:
```diff
@@ -5,6 +5,63 @@
+ * @return string Returns the rendered widget as a string.
+ */
+function render_widget_by_id( $id ) {
+	// Code extracted from src/wp-includes/widgets.php.
```
Let's mention which method this code comes from (`dynamic_sidebar()`, yeah?).

When this code is merged into Core, I'd like to see this part of `dynamic_sidebar()` split out into its own method. We'll need to remember to do this somehow e.g. a Trac ticket or a `TODO` comment here.

noisysocks reviewed (changes requested):
Fundamentally, I think things here are looking really good!

The code that renders a widget is pretty tough to follow (as it is in Core 😅), but I think we can aim to clean a lot of that up if/when this feature is merged into Core e.g. by splitting `dynamic_sidebar()` up.

Before we merge this, I'd really like to see us rethink the REST API design and possibly just ditching it altogether in favour of a regular AJAX action. Happy to chat about this—ping me!

Have you thought about adding E2E tests for widgets? (Can be done as a follow-up.)

jorgefilipecosta committed:
Legacy widget in progress

jorgefilipecosta committed:
Remove all reference widget logic from the endpoint and used ajax admin endpoint instead

jorgefilipecosta commented on lib/class-wp-rest-widget-updater-controller.php:
```diff
@@ -34,13 +34,30 @@ public function __construct() {
 				$this->namespace,
-				// Regex representing a PHP class extracted from http://php.net/manual/en/language.oop5.basic.php.
-				'/' . $this->rest_base . '/(?P<identifier>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)/',
+				'/' . $this->rest_base . '/',
```
Hi @noisysocks, I ended up applying another big change to this PR.
I noticed that if we call the existing ajax admin endpoint without any changes, I retrieve its edit form. So instead of reimplementing this logic for reference widgets, I'm just relying on the endpoint.

For the class widgets, we can not use the endpoint, because there is not a widget instance at all. I kept the rest endpoint but modified it as suggested to be clear of what resource we provide a widget-form.
It may be useful e.g: to build an alternative widget edit UI, we can use this endpoint to retrieve a widget edit form and we can use the server-side render endpoint to retrieve the widget frontend.
I could have made it an ajax endpoint but the code would be more complex, e.g: we get validation and permission checked by the rest API.

Regarding the POST vs GET, the problem with get is that URLs have a max character limit depending on the browser. A widget may have a text area and use lots of character so it seemed safer to use POST. Also, widgets may have legitimate reasons to have side effects e.g: file uploads. Using a get this widgets would break.

noisysocks replied:
> I could have made it an ajax endpoint but the code would be more complex, e.g: we get validation and permission checked by the rest API.

I still have a slight preference for using an AJAX endpoint instead of a REST API endpoint because of how implementation-specific this endpoint is compared to the rest (ha!) of the REST API. Happy to bite my tongue and have us iterate on this, though. I'm glad that the endpoint has been marked experimental 🙂

It might be good to loop in the REST API team at some point on this question.

> Regarding the POST vs GET, the problem with get is that URLs have a max character limit depending on the browser. A widget may have a text area and use lots of character so it seemed safer to use POST. Also, widgets may have legitimate reasons to have side effects e.g: file uploads. Using a get this widgets would break.

Makes sense, thanks 👍

jorgefilipecosta commented:
Hi @noisysocks, I applied a considerable reactor in this PR and addressed most of your suggestions. Please, let me know what are your thoughts on the PR now.

noisysocks commented on packages/block-library/src/legacy-widget/edit/dom-manager.js:
```diff
@@ -61,11 +79,12 @@ class LegacyWidgetEditDomManager extends Component {
-						<input type="hidden" name="widget_number" className="widget_number" value={ widgetNumber } />
-						<input type="hidden" name="multi_number" className="multi_number" value="" />
-						<input type="hidden" name="add_new" className="add_new" value="" />
+						{ isReferenceWidget && ( <><input type="hidden" name="widget-id" className="widget-id" value={ id } />
```
nit: Let's put a newline in between the `<>` and `<input>`.

noisysocks commented on packages/block-library/src/legacy-widget/index.php:
```diff
@@ -60,14 +105,20 @@ function register_block_core_legacy_widget() {
+				'idBase'       => array(
+					'type' => 'string',
+				),
+				'widgetNumber' => array(
```
nit: Is it redundant to prefix all of these attributes with `widget`? The block is called a `legacy-widget`, after all.

noisysocks commented on packages/block-library/src/legacy-widget/edit/handler.js:
```diff
@@ -45,78 +59,135 @@ class LegacyWidgetEditHandler extends Component {
+			formData.append( 'widget-height', '200' );
+			formData.append( 'savewidgets', this.widgetNonce );
+			httpRequest.open( 'POST', window.ajaxurl );
+			const self = this;
```
Let's use a `=>` function instead of a `function()` function below so that `const self = this;` isn't necessary.

noisysocks reviewed (approved):
This is a big improvement—let's get it in so that it gets the testing and iteration that it deserves. Thanks for the hard work, Jorge! 🙂

jorgefilipecosta committed:
Minor code "style" fixes
